8154043: Fields not reachable anymore by tab-key, because of new tabbing behaviour of radio button groups.

ktakakuri opened the pull request:
This is a backport of JDK-8154043: Fields not reachable anymore by tab-key, because of new tabbing behaviour of radio button groups.

Applying the JDK-8154043 fix as is will result in a regression of JDK-8182577.
The fix of JDK-8182577 adds an interface for JDK10, therefore this fix cannot be backported simply for JDK8u.
So, I propose to judge the buttonModel is an instance of DefaultButtonModel.

Testing:
java/awt
javax/swing
ButtonGroupLayoutTraversalTest.java
bug8033699.java
DefaultButtonModelCrashTest.java
on Windows x86_64

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] [JDK-8154043](https://bugs.openjdk.org/browse/JDK-8154043) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))
- [ ] [JDK-8182577](https://bugs.openjdk.org/browse/JDK-8182577) needs maintainer approval

### Issues
 * [JDK-8154043](https://bugs.openjdk.org/browse/JDK-8154043): Fields not reachable anymore by tab-key, because of new tabbing behaviour of radio button groups. (**Bug** - P3 - Rejected)
 * [JDK-8182577](https://bugs.openjdk.org/browse/JDK-8182577): Exception when Tab key moves focus to a JCheckbox with a custom ButtonModel (**Bug** - P3)


### Reviewers
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev.git pull/285/head:pull/285` \
`$ git checkout pull/285`

Update a local copy of the PR: \
`$ git checkout pull/285` \
`$ git pull https://git.openjdk.org/jdk8u-dev.git pull/285/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 285`

View PR using the GUI difftool: \
`$ git pr show -t 285`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/285.diff">https://git.openjdk.org/jdk8u-dev/pull/285.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk8u-dev/pull/285#issuecomment-1588441370)
</details>

ktakakuri commented:
Could someone please review this backport?

ktakakuri commented:
@mrserb 
I issued this PR in relation to https://github.com/openjdk/jdk8u-dev/pull/212.
Could you please review this backport?

mrserb commented:
Sorry I did not have a time to look at this PR, will try to look soon.

ktakakuri commented:
I corrected the copyright date.

jerboaa commented:
As this fix includes the test from JDK-8182577, please use `/isssue add JDK-8182577`.

ktakakuri commented:
/open

mrserb commented:
keep open

jerboaa commented:
> As this fix includes the test from JDK-8182577, please use `/isssue add JDK-8182577`.

Still relevant.

ktakakuri commented:
/open

jerboaa commented:
My bad `/issue add JDK-8182577`

ktakakuri commented:
/issue add JDK-8182577

ktakakuri commented:
@phohensee 
Is it OK to request a fix request?

ktakakuri commented:
@theRealAph 
What is the reason for fix-no?
What should I do next?

ktakakuri commented:
What is the status of this backport now?

phohensee commented:
I filed a backport JBS issue [JDK-8337619](https://bugs.openjdk.org/browse/JDK-8337619), and an associated backport CSR [JDK-8337620](https://bugs.openjdk.org/browse/JDK-8337620). Follow the standard CSR process to get it reviewed and approved. It may not be approved, since it add a public method to an existing API, which afaik can't be done without an update JSR.

ktakakuri commented:
Is there anything I should do now?
Should I just wait for a review?

ktakakuri commented:
Unlike jdk10, this backport does not add any interfaces.
Is it still necessary for the CSR to be approved?

theRealAph commented:
> Unlike jdk10, this backport does not add any interfaces. Is it still necessary for the CSR to be approved?

I don't think so. But there are some complex issues with the jdk 8 backport described in the comments to https://bugs.openjdk.org/browse/JDK-8154043 , which I think is why the 8 bakport was denied in 2016. What I do not understand is why, when the backport was denied 8 years ago, it should be allowed today. If someone can explain that, we can have a meningful discussion.

ktakakuri commented:
Why was the jdk8 backport denied in 2016?
I tried to find it through JBS and email but could not find it.
If adding an interface is the problem, this backport is fine.

ktakakuri commented:
@theRealAph If you have any information, I would appreciate it if you could share it with me.

ktakakuri commented:
I comment to not close.

ktakakuri commented:
I comment to not close.

gnu-andrew reviewed (changes requested):
The change to use `DefaultButtonModel` makes sense as this is the approach used in [JDK-8074883](https://bugs.openjdk.org/browse/JDK-8074883) prior to being replaced in [JDK-8182577](https://bugs.openjdk.org/browse/JDK-8182577)

If we are including the relevant parts of [JDK-8182577](https://bugs.openjdk.org/browse/JDK-8182577) in this change (the test case and the indentation fix to the `member isEnabled` if statement) then we should also include the correction to the indentation of `group.getElements()` and presumably the `instanceof` test?

There is also a closing brace with bad indentation in `ButtonGroupLayoutTraversalTest.java` (line 78). I'm not sure how that crept in.

gnu-andrew commented:
> Why was the jdk8 backport denied in 2016? I tried to find it through JBS and email but could not find it. If adding an interface is the problem, this backport is fine.

There is a [comment](https://bugs.openjdk.org/browse/JDK-8154043?focusedId=13996474&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13996474) on the bug from 2016 that says "No plans to fix for JDK 7 and 8", but it doesn't go into details as to whether this is due to risk or resources, so it's not clear if that is actually a denial or not. Given this was a fix made during the development of 9, the code it was developed against should be fairly close to 8u's and indeed, this backport only really diverges code-wise in that it also incorporates 8182577.

That comment also seems to have been superseded by the backport of this change (and then 8182577) to Oracle's 8u311 by Alexey Ivanov.  I think it's also worth noting that this is a regression from 8 GA due to [JDK-8033699](https://bugs.openjdk.org/browse/JDK-8033699) being introduced in 8u45.  If that backport was suitable for an update release to break this, then presumably the fix should be too.

This brings us back to @theRealAph's question as to why to do this now if it has been broken for so long. What piqued your interest in this bug? Are you or others hitting it in real use? Or was it just a case of Oracle fixed it, so we should?

gnu-andrew commented:
/reviewers 2 reviewer

Bumping to make sure the requested changes are made if we do decide to go ahead with this.

gnu-andrew commented:
/csr unneeded

The associated CSR [JDK-8182695](https://bugs.openjdk.org/browse/JDK-8182695) is for adding the `getModel()` method in [JDK-8182577](https://bugs.openjdk.org/browse/JDK-8182577) but this patch does not add that method (and should not). It is just bringing over the test case and some minor code fixes.

ktakakuri commented:
/open

ktakakuri commented:
@gnu-andrew This issue is impacting several users' applications. In conjunction with the related backport of 8074883, I would appreciate it if this could be integrated until the October 2025 release. We understand you are busy, but would you be able to review this?

ktakakuri commented:
@theRealAph It was under review by @gnu-andrew , but he has not commented since June 6. Would you be possible to take over the review of this pull request  as a maintainer? I'm hoping to get this merged in time for the October release. I have addressed all the comments and suggestions he provided so far. I'm happy to address any new comments and suggestions.

ktakakuri commented:
@jerboaa Would you please review this pull request? I'm hoping to get this merged in time for the October release.
@phohensee Would you please review this pull request again?

ktakakuri commented:
I have merged the latest master branch and confirmed that there are no problems after running :jdk_awt and :jdk_swing tests on Windows.

@jerboaa I previously received feedback from another maintainer on this PR. However, there hasn't been any further response for some time, and I'm wondering if there's an opportunity for it to be reviewed. JDK-8154043's bug is impacting several users' applications.
If your schedule permits, would it be possible for you to take a look at this PR?

Thanks.

gnu-andrew commented:
> > we should also include the correction to the indentation of group.getElements() and presumably the instanceof test?
> 
> Thanks for pointing that out. I've corrected the indentations. What does "the instanceof test" mean? 

I think I just meant this in the 8182577 11u patch:
~~~
-                            if (member.isVisible() && member.isDisplayable() &&
-                                   member.isEnabled() && member.isFocusable()) {
+                            if (member instanceof JToggleButton &&
+                                 member.isVisible() && member.isDisplayable() &&
+                                 member.isEnabled() && member.isFocusable()) {
~~~
Of course, it doesn't have the `instanceof` test in the 8u version, so sorry for the confusion. Anyway, I compared the 11u version of `LayoutFocusTraversalPolicy.java` with your current patched 8u version and it looks good, so you may have just sorted this one out from the start.

gnu-andrew commented:
> In bug8033699.java, line 2, leave the end copyright date at 2019. Other than that, lgtm.
> 
> Also, please enable GHA on this repo.

Good catch. 8226892 was already backported with a later copyright date on this file.

ktakakuri commented:
@gnu-andrew Thank you for your review. I'm planning to reopen and work on https://github.com/openjdk/jdk8u-dev/pull/212  related to this PR soon, so I'd appreciate it if you could review that as well when the time comes.

ktakakuri commented:
/approval JDK-8154043 cancel

ktakakuri commented:
@gnu-andrew JDK-8154043 has already been labeled with jdk8u-fix-no, and I can't submit the request again. I want to resubmit it. How should I proceed?

gnu-andrew commented:
> @gnu-andrew JDK-8154043 has already been labeled with jdk8u-fix-no, and I can't submit the request again. I want to resubmit it. How should I proceed?

I've approved manually on the bugs. This should be good to go.

ktakakuri commented:
@gnu-andrew Thank you for your prompt support.

ktakakuri commented:
/integrate
